scheduler/src/cook/mesos/mesos_mock.clj

```diff
@@ -0,0 +1,454 @@
+
+(defn complete-task!
+  "Marks the task as complete in the state and updates the resources
+   Returns the updated state"
+  [state task-id scheduler driver task-state]
+  (log/debug "Completing task" {:task-id task-id :state task-state :task (get-in state [:task-id->task task-id])})
+  (if-let [task (get-in state [:task-id->task task-id])]
+    (do
+      (.statusUpdate scheduler driver (mesos-type/->pb :TaskStatus 
+                                                       {:task-id {:value task-id}
+                                                        :state task-state}))
+      (-> state
+          (update :task-id->task #(dissoc % task-id))
+          (update :task-id->completed-task #(assoc % task-id (assoc task :complete-time (t/now))))
+          (update-in [:slave-id->host (:value (:slave-id task)) :available-resources] 
+                     #(combine-resources (:resources task) %))))
+    state))
+
+(defmulti handle-action!
+  "Handles the particular action, likely calling methods on the scheduler
+   and returning an updated state
+
+   Parameters:
+   `action` the option to dispatch off
+   `data` data corresponding to the action and specific to the action
+   `state` the current state of 'mesos'
+   `driver` the driver used by the scheduler
+   `scheduler` the scheduler framework connected to mesos"
+  (fn [action data state driver scheduler]
+    action))
+
+(defmethod handle-action! :decline
+  [action offer-id {:keys [offer-id->offer task-id->task slave-id->host] :as state} driver scheduler]
+  (if-let [offer (get offer-id->offer offer-id)]
+    (-> state
+        (update :offer-id->offer #(dissoc % offer-id))
+        (update-in [:slave-id->host (:value (:slave-id offer)) :available-resources]
+                   #(combine-resources (:resources offer) %)))
+    (throw (ex-info "Unknown offer-id" {:offer-id offer-id}))))
+
+(defmethod handle-action! :kill-task
+  [action task-id {:keys [offer-id->offer task-id->task slave-id->host] :as state} driver scheduler]
+  (log/debug "Killing task " {:task-id task-id})
+  (complete-task! state task-id scheduler driver :task-killed))
+
+
+(defmethod handle-action! :launch
+  [action {:keys [offer-ids tasks] :as in} {:keys [offer-id->offer task-id->task slave-id->host] :as state} driver scheduler]
+  (log/debug "In launch handle-action!" in)
+  (let [offers (map offer-id->offer offer-ids)
+        slave-id (:value (:slave-id (first tasks)))]
+    (when-not (every? #(= slave-id (:value (:slave-id %))) tasks)
+      (throw (ex-info "All tasks must have same slave-id" 
+                      {:slave-ids (map (comp :value :slave-id) tasks)})))
+    (when-not (get slave-id->host slave-id)
+      (throw (ex-info "Unknown slave-id" {:slave-id slave-id :known-slave-ids (keys slave-id->host)})))
+    (when (some nil? offers) 
+      (throw (ex-info "Unknown offers" 
+                      {:unknown-offers 
+                       (remove #(contains? (set (keys offer-id->offer))) 
```

Comment:
Yes, thanks for catching that. I added these checks to save me grief when there was a bug upstream but I guess I haven't exercised that one yet.